pkg/controller/seed-controller-manager/constraint-controller/controller.go

```diff
@@ -235,29 +263,51 @@ func (r *reconciler) reconcile(ctx context.Context, constraint *kubermaticv1.Con
 		}
 	}
 
-	// constraint creation
+	if err = r.cleanupConstraint(ctx, log, constraint, undesiredClusterList); err != nil {
+		return err
+	}
+
+	if err = r.createConstraint(ctx, log, constraint, desiredClusterList); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (r *reconciler) createConstraint(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint, clusterList *kubermaticv1.ClusterList) error {
 	constraintCreatorGetters := []reconciling.NamedKubermaticV1ConstraintCreatorGetter{
 		constraintCreatorGetter(constraint),
 	}
 
-	return r.syncAllClustersNS(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
+	if err := r.syncAllClustersNS(ctx, log, constraint, clusterList, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
 		return reconciling.ReconcileKubermaticV1Constraints(ctx, constraintCreatorGetters, namespace, seedClient)
-	})
+	}); err != nil {
+		return err
+	}
+
+	return nil
 }
 
-func (r *reconciler) cleanupConstraint(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint) error {
+func (r *reconciler) cleanupConstraint(ctx context.Context, log *zap.SugaredLogger, constraint *kubermaticv1.Constraint, clusterList *kubermaticv1.ClusterList) error {
 
-	if err := r.syncAllClustersNS(ctx, log, constraint, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
+	if err := r.syncAllClustersNS(ctx, log, constraint, clusterList, func(seedClient ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint, namespace string) error {
 
 		log := log.With("constraint", constraint)
 		log.Debugw("cleanup processing:", namespace)
 
-		err := seedClient.Delete(ctx, &kubermaticv1.Constraint{
+		constraint = &kubermaticv1.Constraint{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:      constraint.Name,
 				Namespace: namespace,
 			},
-		})
+		}
+
+		err := seedClient.Get(ctx, types.NamespacedName{Name: constraint.Name, Namespace: namespace}, constraint)
+		if err != nil && errors.IsNotFound(err) {
+			return nil
+		}
+
+		err = seedClient.Delete(ctx, constraint)
```

Comment:
I think just delete is fine, if its not there we get the same result.

Reply:
This is to avoid doing API calls when is not needed, the `Get` is cached while delete will hit KAS.

Reply:
ahh good point

Reply:
@harshita-sharma012 a comment is welcome as this is not obvious at first glance

Reply:
@lsviben do we need this part of code as with the new changes you suggested for filtering we are not getting any unnecessary clusters